### PR TITLE
[New Model] Support MOSS-TTS-LOCAL

### PR DESCRIPTION
**Pull request by sharonlu2**
## Purpose

This PR adds support for the MOSS-TTS-Local model in vLLM-Omni, enabling local-transformer TTS inference with Stage 0 RVQ-code generation and Stage 1 codec decoding.

Key changes:
- Integrate MOSS-TTS-Local model executor in `vllm_omni/model_executor/models/moss_tts/`.
- Add sync/async stage configs for offline inference.
- Include offline inference examples under `examples/offline_inference/moss_tts_local/`.
- Add focused tests for MOSS local stage input processing, AR-stage sampling/FSM behavior, decoder handling, registry aliasing, and connector metadata forwarding.

## Test Plan

- `pytest tests/model_executor/models/moss_tts tests/model_executor/stage_input_processors/test_moss_tts_local.py tests/distributed/omni_connectors/test_chunk_transfer_adapter.py tests/worker/test_omni_gpu_model_runner.py -q`
- E2E smoke test with downloaded `OpenMOSS-Team/MOSS-TTS-Local-Transformer` and `OpenMOSS-Team/MOSS-Audio-Tokenizer`:

```bash
MOSS_AUDIO_TOKENIZER_PATH=/path/to/moss-audio-tokenizer \
CUDA_VISIBLE_DEVICES=0 \
python examples/offline_inference/moss_tts_local/end2end.py \
    --model /path/to/moss-tts-local \
    --stage-configs-path vllm_omni/model_executor/stage_configs/moss_tts_local.yaml \
    --text "Hello world." \
    --output-dir ./out
```

## Test Result

- `python -m pytest tests/model_executor/models/moss_tts tests/model_executor/stage_input_processors/test_moss_tts_local.py tests/distributed/omni_connectors/test_chunk_transfer_adapter.py tests/worker/test_omni_gpu_model_runner.py -q`
- Result: `38 passed in 0.71s`
- Performance numbers should be re-run after this branch is validated on the target GPU environment.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/).
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and examples for a new model. **Please run `mkdocs serve` to sync the documentation editions to docs.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/models/registry.py [hunk 1]
```diff
@@ -118,6 +118,16 @@
         "mimo_audio_code2wav",
         "MiMoAudioToken2WavForConditionalGenerationVLLM",
     ),
+    "moss_tts_delay": (
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/registry.py:121]: [blocking] `("moss_tts", "moss_tts", "MossTTSForConditionalGeneration")` points to `vllm_omni/model_executor/models/moss_tts/moss_tts.py`, which is not in this PR. HF auto-routing on `model_type=moss_tts_delay` (the value in the HF config of OpenMOSS-Team/MOSS-TTS-Local-Transformer) hits this entry first and will raise `ModuleNotFoundError`. Only the `MossTTSForConditionalGeneration` entry below works, and that one only because the YAML overrides `model_arch`. Either point this entry at `moss_tts_local` or drop it.

File: vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py [hunk 2]
```diff
@@ -160,11 +160,13 @@ def _poll_single_request(self, request: Request):
 
                 new_ids = payload_data.get("code_predictor_codes", [])
                 request.prompt_token_ids = new_ids
-                # Pass additional fields (like left_context_size) to the request
-                # Only pass chunk context metadata in additional_information
-                request.additional_information = {}
-                if "left_context_size" in payload_data:
-                    request.additional_information["left_context_size"] = payload_data["left_context_size"]
+                # Testing-only: preserve chunk metadata end-to-end so Stage 1
```
Comment (linyueqian) [inline vllm_omni/distributed/omni_connectors/transfer_adapter/chunk_transfer_adapter.py]: [blocking] The previous code forwarded only `left_context_size`; the new code forwards every key except `code_predictor_codes`. This change applies to every model that uses `SharedMemoryConnector`, not just MOSS, so downstream consumers may see fields they were not designed for. The "Testing-only" prefix in the comment is itself a signal this should not land. Please scope the change to MOSS (for example with a model-specific allow-list keyed by the YAML) or revert this hunk.

File: vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py [hunk 3]
```diff
@@ -0,0 +1,1056 @@
...
+                    else:
+                        gen_keep = row[self.gen_slot_id].clone()
+                        end_keep = row[self.audio_end_id].clone()
+                        row.fill_(neg_inf)
+                        row[self.gen_slot_id]  = gen_keep
+                        row[self.audio_end_id] = end_keep
+                else:
+                    if self.pad_token_id >= 0:
+                        row[self.pad_token_id] = neg_inf
+                    row[self.gen_slot_id]  = neg_inf
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py]: [blocking] FSM dead-lock: while `is_audio=False`, `compute_logits` masks `gen_slot_id` and `audio_end_id` here, so the only entry to audio mode is sampling `audio_start_token_id`. On the prefill call `_last_request_ids` is empty (it is assigned later in `forward` at line 607), so `compute_logits` early-returns at line 917 without masking and the sampler picks an arbitrary text token. `is_audio` then stays `False` forever, `_local_forward` is never invoked, and the pipeline hangs waiting for codes. Reproduced on h20-server-1 with both sync and async YAMLs; full repro in the overview.

**Review by linyueqian — COMMENTED** [review]
Thanks for the PR. I read through the code and tried to run `end2end.py` on h20-server-1 (8x H20 141GB, vllm 0.20.0, vllm-omni current main). The pipeline does not generate audio in this environment: stage 0 hangs after the first request and stage 1 waits for output that never arrives. Root cause is in the inline comment at `moss_tts_local_ar_stage.py:949` (FSM dead-lock at prefill time, compounded by `_extract_request_info` silently swallowing the underlying error at line 787).

Other items I would like addressed before this can land:

1. `HPML-metrics/` adds about 9.6K lines of Jupyter notebooks, vast_ai notes, and scratch throughput scripts that reference filenames like `moss_tts_delay.yaml` that are not in this PR. Please drop the directory or move it to a separate PR.
2. The `"moss_tts_delay"` entry in `models/registry.py` is dead (no `moss_tts/moss_tts.py` in the diff). Please point it to `moss_tts_local` or remove it.
3. The `chunk_transfer_adapter.py` change is labeled "Testing-only" in its own comment and loosens the metadata contract for every model that uses `SharedMemoryConnector`, not just MOSS. See inline.
4. The PR test plan promises `tests/model_executor/test_moss_tts_local.py` but no test file is in the diff. Please add at least a smoke test that exercises stage 0 prefill, decode, and stage 1 codec on one prompt.
5. PR currently has merge conflicts against main; please rebase.

Performance verdict: I cannot reproduce the 1.8x throughput / 39.2x FCL numbers in the PR description, because the pipeline does not generate audio at all on current main. Once the FSM blocker is fixed and `enforce_eager` is revisited (both YAMLs ship `enforce_eager: true`, which disables `torch.compile` and CUDA graphs), please re-run with a fresh git SHA and update the description.

Repro (after fetching the PR branch and downloading `OpenMOSS-Team/MOSS-TTS-Local-Transformer` + `OpenMOSS-Team/MOSS-Audio-Tokenizer`):

```
MOSS_AUDIO_TOKENIZER_PATH=/path/to/moss-audio-tokenizer \
CUDA_VISIBLE_DEVICES=0 \
python end2end.py \
    --model /path/to/moss-tts-local \
    --stage-configs-path .../moss_tts_local.yaml \
    --text "Hello world." \
    --output-dir ./out
```

Result: 600s timeout, no wav written. Same with `moss_tts_local_async.yaml`, with `MOSS_TTS_LOCAL_KV_CACHE=0`, and after a local patch that also reads `model_intermediate_buffer` in the AR stage. The `[MossTTS Local KV]` debug log from `_local_forward` never fires, confirming that path is not entered.

File: vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py [hunk 4]
```diff
@@ -0,0 +1,1056 @@
...
+                return [], [], []
+            if isinstance(attn_meta_dict, list):
+                attn_meta_dict = attn_meta_dict[0]
+            attn_meta = next(iter(attn_meta_dict.values()))
+            qsl = attn_meta.query_start_loc.cpu().tolist()
+            num_reqs = len(qsl) - 1
+            seq_lens = [qsl[i + 1] - qsl[i] for i in range(num_reqs)]
+            decode_positions = [qsl[i] for i, s in enumerate(seq_lens) if s == 1]
+        except Exception as exc:
+            logger.debug("[MossTTS AR] _extract_request_info failed: %s", exc)
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py]: [important] `logger.debug` plus a bare `except Exception` quietly hides every cause of failure in `_extract_request_info`. Combined with the FSM issue above, this is why the hang has no visible error in logs. Promote to `logger.warning` with `exc_info=True`, or at minimum log the repr.

File: vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py [hunk 5]
```diff
@@ -0,0 +1,1056 @@
...
+    # =======================================================================================
+
+    @torch.no_grad()
+    def _local_forward(
+        self,
+        global_hidden: torch.Tensor, # [B, D_global]  (B = num decode seqs)
+        forced_text_token_id: int,
+        temperature: float = 1.0,
+        top_k: int = 50,
+        top_p: float = 0.95,
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py:678]: [important] `top_p` is in the signature and the AR YAML defaults (`top_p: 0.95`), but the function body only applies `top_k` filtering. The parameter is currently a no-op. Either implement nucleus filtering or drop the parameter so callers do not get false sampling control.

File: vllm_omni/model_executor/stage_configs/moss_tts_local.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      dtype: bfloat16
+      max_num_seqs: 1
+      model_stage: ar_stage
+      model_arch: MossTTSForConditionalGeneration
+      worker_type: ar
+      scheduler_cls: vllm_omni.core.sched.omni_ar_scheduler.OmniARScheduler
+      tensor_parallel_size: 1 # scale up for multi-GPU
+      gpu_memory_utilization: 0.3
+
+      enforce_eager: true
```
Comment (linyueqian) [inline vllm_omni/model_executor/stage_configs/moss_tts_local.yaml:21]: [important] `enforce_eager: true` is set on both stages of both YAMLs (`moss_tts_local.yaml:22`, `:57`, `moss_tts_local_async.yaml:21`, `:60`). This disables `torch.compile` and CUDA graphs. The 1.8x throughput / 39.2x FCL numbers in the PR description cannot come from this config. Please document the exact YAMLs and SHA that produced the benchmarks, or drop `enforce_eager` if the model is graph-safe.

File: vllm_omni/model_executor/models/moss_tts/moss_tts_local_decoder.py [hunk 7]
```diff
@@ -0,0 +1,648 @@
...
+
+    def _enter_streaming(self, request_id: str) -> None:
+        if request_id in self._streaming_states:
+            return
+        stack = ExitStack()
+        codec = self._codec.codec
+        for decoder_module in codec.decoder:
+            if hasattr(decoder_module, "streaming") and callable(decoder_module.streaming):
+                stack.enter_context(decoder_module.streaming(batch_size=1))
+        self._streaming_states[request_id] = stack
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts_local_decoder.py:341]: [important] `_streaming_states[request_id] = stack` is only popped on `is_finished=True` or when the request key rotates in batched mode. Aborted requests (preemption, OOM, scheduler eviction) leave the `ExitStack` open and the codec KV state stuck. The same applies to `transfer_manager._moss_frames_sent` in the input processor. Hooking into the engine's abort callback would be cleaner.

File: vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py [hunk 8]
```diff
@@ -0,0 +1,1056 @@
...
+        attn_metadata                                  = None,
+        intermediate_tensors: Optional[IntermediateTensors] = None,
+        inputs_embeds: Optional[torch.Tensor]          = None,
+        **kwargs,
+    ) -> OmniOutput:
+        with record_function("stage0/forward"), _TIMER.gpu("stage0/forward_total"):
+            # Step 0: obtain information from vLLM context
+            with record_function("stage0/extract_info"), _TIMER.cpu("stage0/extract_info"):
+                request_ids, seq_lens_per_req, decode_positions = self._extract_request_info(
+                    kwargs.get("runtime_additional_information"),
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/moss_tts/moss_tts_local_ar_stage.py]: [suggestion] The decoder at `moss_tts_local_decoder.py:475` already falls back from `model_intermediate_buffer` to `runtime_additional_information`. Mirror that here, since vllm-omni 0.20 emits a deprecation warning recommending the new key.

**Comment by gcanlin** [comment]
Sorry, I didn't notice this PR before. Are you interested in https://github.com/vllm-project/vllm-omni/issues/4676, which is still continuing to optimizing the performance for MOSS-TTS-Local.

**Comment by yscout** [comment]
> Sorry, I didn't notice this PR before. Are you interested in #4676, which is still continuing to optimizing the performance for MOSS-TTS-Local.

Yeah sure we can do that.
